KEP-29: Operator dependencies (#1454)
kudobuilder/kudo: keps/0029-operator-dependencies.md

      * [Client-Side](#client-side)
      * [Server-Side](#server-side)
      * [Parameterization](#dependencies-parameterization)
    * [Update](#update)
    * [Upgrade](#upgrade)
    * [Uninstalling](#uninstalling)
  * [Risks and Mitigation](#risks-and-mitigation)
* [Alternatives](#alternatives)

## Summary

This KEP aims to improve operator user and developer experience by introducing operator dependencies.

## Motivation

Recent operator development has shown that complex operators often depend on other operators to function. One workaround commonly seen can be found in the [Flink Demo](https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud). It requires manual installation of the dependency operators while skipping the instance creation (using the `--skip-instance` option). The `Instance` is then "manually" created during the execution of the `deploy` plan. This KEP is aiming at improving this experience, streamlining the developer and user experience.

### Goals

Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only.

### Non-Goals

> Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded and the new version is incompatible with the current Kafka `Instance`? How can compatibility be ensured? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.

Reviewer: Maybe we should add that using the same dependency instance from two points of the dependency graph is not supported? As written below, if we have two operators that wants to install zookeeper, we need to create two zk instances and can't use a single one?

Author: > Maybe we should add that using the same dependency instance from two points of the dependency graph is not supported?

It is **not** as it will create a dependency circle.

The instance controller (IC) then traverses the dependency graph in the evaluation order `O` and executes each vertex. Practically this means creating a corresponding `Instance` resource (`Operator` and `OperatorVersion` already exist) and wait for it to become healthy, meaning that its `deploy` plan was executed successfully. KUDO manager already has [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented. We would additionally add the top-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of each dependency `Instance` (e.g. `BB`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).

The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.

##### Dependencies Parameterization

> First, we need to provide parameters to different operators separately. For individual parameters (`-p` option) we can use namespaced names e.g. `-p <operatorName>.<key>=<value>`. For the parameter files (`--parameterFile` option) we could extend `parameterFile` schema (which is currently a simple map) with top-level fields e.g.:

Reviewer: Is that enough? I'm thinking about having multiple instances of the same Operator:
For example a top-level operator requires a) Zookeeper, and b) Kafka, which itself requires zookeeper as well.

In that case we might want to specify different amounts of nodes for each zookeeper, right? I guess we could also allow to use the same zk instance, but that would probably make things a lot more difficult?

Reviewer: Additionally, in KEP-28 there is an idea to have plan-specific parameters which might be prefixed with the plan names, so this might clash there. 

Thinking about it, maybe it would work if we'd use the "task" name as a prefix, and make task-specific parameters instead of plan specific ones?

Author: >  I'm thinking about having multiple instances of the same Operator:
For example a top-level operator requires a) Zookeeper, and b) Kafka, which itself requires zookeeper as well.

I've been thinking about it - what does it actually mean, to depend on **an already existing* `Instance`?  

a) **resource referencing:** for an existing Zookeeper instance `zook`, you'd most probably like to reference it in e.g. Kafkas `ZOOKEEPER_URI: zook-zookeeper-0.zook-hs:2181`. Since we don't support terraform style `Instance` resource referencing you could simply "hardcode" it.
b) **installation check:** make sure that the `zook` instance is running healthy. Not yet supported.

But otherwise, you're right - reusing the same `Instance` during deployment is not supported as it would create circular dependency. I will add this to the KEP.

Author: > Additionally, in KEP-28 there is an idea to have plan-specific parameters which might be prefixed with the plan names, so this might clash there.

I hope we would go with proposal nr. 5 and avoid _"plan-specific parameters"_

```yaml
apiVersion: v1beta1
operator: zookeeper
parameters:
  foo: bar # other parameterFile keys and values
```

Second, parameters are normally parsed to the CLI and persisted in the `Instance` resource. Since the creation of the instances is delayed we need a place to store the parameters until then. We could keep them in the task definition e.g.:

```yaml
tasks:
- name: demo
  kind: Operator
  spec:
    ...
    parameters: # optional, a map of parameter keys and values constructed by expanding the parameterFiles along with individual parameters

```

KUDO manager would then read them and populate `Instance.Spec.Parameters` as usual. The drawback here is that this would likely stretch currently slim task definitions.

> Alternatively, we could introduce a new field e.g. `OperatorVersionSpec.InstallationParameters` alongside with the existing [Parameters](https://github.com/kudobuilder/kudo/blob/master/pkg/apis/kudo/v1beta1/operatorversion_types.go#L35). Both approaches contradict our current data model where such parameters would exist in the `Instance` resource.

Reviewer: Saving this in the OperatorVersionSpec would also clash with - as mentioned above - multiple instances that need to be installed.

Author: I've added a clarification in the KEP 👍

## Proposal

KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.

### Implementation details

#### Operator Task

A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:

```yaml
apiVersion: kudo.dev/v1beta1
name: "kafka"
operatorVersion: "1.3.1"
kudoVersion: 0.12.0
kubernetesVersion: 1.14.8
appVersion: 2.5.0
url: https://kafka.apache.org/
tasks:

>   - name: zookeeper-operator

Reviewer 2: so if kafka wants to target ZK via connection string, operator has to take care of that via params, right? that is also not in scope of this

Author: that would be resource referencing and/or output variables. out of scope for this KEP.

    kind: Operator

>     spec:

Author: It's not just the `Instance`, it's the whole operator package, right? E.g. `spec.package` is the name of the operator package and not the instance. I think `kind: Operator` is more fitting.

>     spec:

Reviewer 3: From my PoV, after seeing `kind: Operator` I thought it only deals with the "need to manually run `kudo install --skip-instance`" issue. It then surprised me that this does not *only* install the operator(version) but also installs the instance as well. :shrug:

>     spec:

Reviewer 3: An alternative I came up with would be no call this `Kudo`, since it's not any operator, but specifically KUDO operators only.

        package: zookeeper

  ...

plans:
  deploy:
    strategy: serial
    phases:
      - name: deploy-zookeeper
        strategy: serial
        steps:
          - name: dummy
            tasks:
              - zookeeper-operator
      - name: deploy-kafka
        strategy: serial
        steps:
          ...
```

> The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:

Reviewer 2: what if instance with that name alrrady exists? Will it continue installing or fail? If it continues, then it's basically reusing already running instance which is a non-goal I understand...

Author: > what if instance with that name alrrady exists? 

this can't happen, right? you can't have two instances [with the same name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/), the same way you can't have two pods with the same name (or is it per namespace? 🤔)

Reviewer 3: @zen-dog it depends on whether you're using an idempotent `create-or-update` or non-idempotent `create-or-fail`, right?

Author: For `Instance`s we use `create-or-fail`, otherwise, we risk changing an already existing instance for some other operator. These problem falls into _"depending on an existing Instance"_ category and is out of scope for this KEP.

```yaml
tasks:
- name: demo

>   kind: Operator

Reviewer 3: ```suggestion
  kind: Instance
```

  spec:
    package: # required, either repo package name, local package folder or an URL to package tarball
    repo: # optional, name of local repository configuration to use

>     appVersion: # optional, a specific app version in the official repo, defaults to the most recent one

Reviewer 2: what if both appVersion and operatorVersion are set and they don't map to existing package? I assume it fails in installation, right? will this be checked before anything is applied to cluster so we don't end up with abandoned OVs?

Author: normally, we would be able to just reuse existing `O/OV` (that's what they are for, right?). but this KEP suggests persisting dependency `Instance` parameters in the `OV` which breaks the conceptual separation between `OV` and `I` 😞 Overall, this is IMHO the biggest downsides of this KEP and I haven't come up with a satisfactory solution yet.

Author: Here a different solution: 

We pre-create `Instances` on the **client-side** but mark them as _Inactive_ with some annotation e.g. `kudo.de/instance-inactive`. To make things easier IAC would only allow this annotation to be set when `CREATE`ing the instance. This annotation would mark a new life-cycle Instance state where it would effectively be ignored by instance controller.

On the **server-side** we would do the same steps described in this KEP but instead of creating `Instances`, we would activate them (by removing the annotation).

Wdyt?

Reviewer 3: @zen-dog I think @alenkacz is asking what happens if a `kafka` OV says it needs an instance of `zookeeper` version 5, but a package for zookeeper 5 is nowhere to be found in the repo. I think it should just fail at CLI stage, since it's not different from trying to install a non-existent operator `badger`.

I don't understand how reusing existing `O/OV` helps with that?

Author: > but a package for zookeeper 5 is nowhere to be found in the repo.

It will fail the same way it fails today for a single operator package with `Error: failed to resolve operator package`

#### Deployment

##### Client-Side

Upon execution of `kudo install ...` command with the above operator definition, CLI will:

1. Collect all operator dependencies by analyzing the `deploy` plan of the top-level operator and compose a list of all operator dependencies (tasks with the kind `Operator`) including **transitive dependencies**
2. Install all collected operators **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order
3. Proceed with the installation of the top-level operator as usual (create `Operator`, `OperatorVersion` and `Instance` resources)

Since we do this step on the client-side we have access to the full functionality of the `install` command including installing operators from the file system. This will come very handy during the development and debugging which arguably becomes more complex with dependencies.

##### Server-Side

Upon receiving a new operator Instance with dependencies KUDO mangers workflow engine will:

> 1. Build a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) by transitively expanding top-level `deploy` plan using operator-tasks as vertices and their execution order (`a` needs `b` to be installed first) as edges

Reviewer 3: So you cannot have operator deps in plans other than deploy? Not sure if this is a problem, but this limitation should probably be mentioned somewhere.
Or perhaps there should be another separate special "dependencies" plan, or this should walk the list of all defined tasks... just wondering..

Author: No, the later sections show dependencies for e.g. `upgrade` and `update` plans. I'm not sure if it is meaningful to allow operator-task outside of these plans but theoretically, it is possible 🤷

The instance controller (IC) then traverses the dependency graph in the evaluation order `O` and executes each vertex. Practically this means creating a corresponding `Instance` resource (`Operator` and `OperatorVersion` already exist) and wait for it to become healthy, meaning that its `deploy` plan was executed successfully. KUDO manager already has [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented. We would additionally add the top-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of each dependency `Instance` (e.g. `BB`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).

The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.

> Note that in the above example could not depend on the same `Instance` as it will create a dependency cycle. So if e.g. `EE` and `CC` are instances of the same `OperatorVersion`, they must use distinct names and will be installed as two separate `Instances`.

Reviewer 3: ```suggestion
Note that in the above example SUBJECT_MISSING_HERE could not depend on the same `Instance` as it will create a dependency cycle. So if e.g. `EE` and `CC` are instances of the same `OperatorVersion`, they must use distinct names and will be installed as two separate `Instances`.
```

tasks:
- name: demo
  kind: Operator
  spec:
    package: # required, either repo package name, local package folder or an URL to package tarball
    repo: # optional, name of local repository configuration to use
    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
    operatorVersion: # optional, a specific operator version in the official repo, defaults to the most recent one
    instanceName: # optional, the instance name
```

As you can see, this closely mimics the `kudo install` CLI command [options](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/cmd/install.go#L56) because at the end the latter will be executed to install the operator. We omit `parameters` and `parameterFile` options at this point as they are discussed in detail [below](#dependencies-parametrization).

#### Deployment

##### Client-Side

Upon execution of `kudo install ...` command with the above operator definition, CLI will:

1. Collect all operator dependencies by analyzing the `deploy` plan of the top-level operator and compose a list of all operator dependencies (tasks with the kind `Operator`) including **transitive dependencies**

> 2. Install all collected operators **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order

Reviewer 3: ```suggestion
2. Install all collected operator versions **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order
```

Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:

```text
AA
├── BB
│   ├── EE
│   │   ├── H
│   │   └── I
│   ├── F
│   └── GG
│       ├── J
│       └── K
├── CC
│   ├── L
│   └── M
└── D

Legend:
- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
- other tasks are marked with single letters e.g. 'D'
- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
```

In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.

> The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` when then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.

Reviewer 3: ```suggestion
The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.
```

# /apis/kudo.dev/v1beta1/namespaces/default/instances/instance-aa
spec:
  parameters:
    BB_PASSWORD: secret
```

During the execution of the `deploy-bb` task, the `bb-params.yaml` is expanded the same way we expand templates during the apply-task execution. The `deploy-bb` operator-task then creates the `Instance-BB` resource and saves the expanded parameter `PASSWORD: secret` in it.

What happens if we have a deeper nested operator-tasks tree e.g.:
```yaml
AA
└── BB
    └── CC
        └── DD  
            └── EE
```
and it is the low-level `EE` operator that needs the password? It is like the dependency injection through constructor parameters: every higher-level operator has to encapsulate the password parameter so that `AA` has the `BB_PASSWORD`, `BB` the `CC_PASSWORD` and so on.

 #### Update

Updating parameters work the same way as deploying the operator. In most cases, the `deploy` plan is executed. Since all dependencies already exist, the KUDO manager will traverse the dependency graph, updating Instance parameters. This will then trigger a corresponding `deploy` plan on each affected `Instance`. If the `Instance` hasn't changed no plan will be triggered.

#### Upgrade

> While an out-of-band upgrade of the individual dependency operators is possible (and practically impossible to prohibit), operators, in general, should be upgraded as a whole to preserve compatibility between all dependencies. An `upgrade` plan execution is very similar to the `deploy` plan. CLI creates new `OperatorVersion` resources for all new dependency operator versions. KUDO manager builds a dependency graph by traversing the `upgrade` plans of the operators and executes them in a similar fashion.

Reviewer 3: ```suggestion
While an out-of-band upgrade of the individual dependency operators is possible (and practically impossible to prohibit until KUDO learns drift detection), operators, in general, should be upgraded as a whole to preserve compatibility between all dependencies. An `upgrade` plan execution is very similar to the `deploy` plan. CLI creates new `OperatorVersion` resources for all new dependency operator versions. KUDO manager builds a dependency graph by traversing the `upgrade` plans of the operators and executes them in a similar fashion.
```

### Implementation details

#### Operator Task

A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:

```yaml
apiVersion: kudo.dev/v1beta1
name: "kafka"
operatorVersion: "1.3.1"
kudoVersion: 0.12.0
kubernetesVersion: 1.14.8
appVersion: 2.5.0
url: https://kafka.apache.org/
tasks:
  - name: zookeeper-operator
    kind: Operator
    spec:
        package: zookeeper

  ...

plans:
  deploy:
    strategy: serial
    phases:
      - name: deploy-zookeeper
        strategy: serial
        steps:
          - name: zookeeper
            tasks:
              - zookeeper-operator
      - name: deploy-kafka
        strategy: serial
        steps:
          ...
```

The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:

```yaml
tasks:
- name: demo

>   kind: Operator

Reviewer 4: We should talk about - but add it as a non-goal to this KEP - to support other Operator tools at this time. KOI KEP still needs to be made.

  spec:
    package: # required, either repo package name, local package folder or an URL to package tarball

>     repo: # optional, name of local repository configuration to use

Reviewer: I'm not sure if it's a good idea to include the repo URL in the operator definition. This would open up a lot of ways for people to sneak in unintended operators and have security implications.

If (as an operator user) want to use only operators from `https://official-operators.kudo.dev`, I don't expect a nested operator from `https://unknown-third-party-operator.com/` to be installed.

Additionally, I would expect problems with air-gapped deployments if we allow this.

This should at least be disabled by default and only enabled via an explicit command line flag if we allow this.

It would be nice to have for local development though.

>     repo: # optional, name of local repository configuration to use

Reviewer 3: > If I (as an operator user) want to use only operators from https://official-operators.kudo.dev, I don't expect a nested operator from https://unknown-third-party-operator.com/ to be installed.

I think it's up to maintainers of `official-operators` to check/enforce that they do not depend on unofficial ones?

> repo names are client controlled and not really standardized

good point, a local repo name does not make any sense in this context

How about just removing this field for now, requiring that the dependent operator+version be available "somehow" to the CLI?

>     repo: # optional, name of local repository configuration to use

Author: > How about just removing this field for now, requiring that the dependent operator+version be available "somehow" to the CLI?

Agree. Initially, I just copied that from the `kudo install` settings but now, I can see, how repo names can be arbitrary and non-standardized between environments. A repo-url might make more sense but let's tackle this later.

    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
    operatorVersion: # optional, a specific operator version in the official repo, defaults to the most recent one

>     instanceName: # optional, the instance name

Reviewer: Good point, It might be valuable to allow prefixes or some other linkage to the group. If I name my top-level operator `abc`, the dependency might/should have the name `abc-zookeeper`

>     instanceName: # optional, the instance name

Author: But that's at the hand of the operator developer, right? If you want, you can name your instances accordingly. If not, KUDO will take the default name, like it does today.

>     instanceName: # optional, the instance name

Reviewer: Then we either have to extend the defaulting (i.e. if the instance name `zookeeper-instance` is already taken, use `zookeeper-instance1` or provide a way for the operator user to configure this.

Otherwise we run into the problem of having multiple operators that use the same dependency:

Operator AA uses zk and doesn't define an instanceName. zk gets installed as `zookeeper-instance`.
Operator BB uses zk and doesn't define an instanceName. zk wants to install as `zookeeper-instance` but it's already there and it fails. The user now has no way to install Operator BB. Or am I missing something?

Making the instanceName required may help, but... Not sure if that's the best way.

>     instanceName: # optional, the instance name

Reviewer: Additionally - If i want to install AA twice, with different instance names, I can do that for the top-level operator. But if the dependencies don't rely on the top-level-instance name, they will always clash.

>     instanceName: # optional, the instance name

Author: These are all good points. Wdyt about making dependency instance names hierarchical? E.g. if the top-level operator instance is named `kafka-prod` then it's ZK instance is named `kakfa-prod.zk-instance`? This way, it is also immediately clear in the `$ k get instances` where the instance belongs to?

>     instanceName: # optional, the instance name

Reviewer: Yeah, I'd say that'd be a good idea. Not sure which separator we should use. `.` vs `-` vs `_`, but some kind of hierarchy would be good.

>     instanceName: # optional, the instance name

Reviewer 3: Would be nice, but I'm not sure the schema for `metadata.name` will like the dots/underscores or the resulting long names in case of deeply nested operators...

>     instanceName: # optional, the instance name

Author: From the ̀[docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names) an object name should:

- contain no more than 253 characters
- contain only lowercase alphanumeric characters, ‘-’ or ‘.’
- start with an alphanumeric character
- end with an alphanumeric character

Dots are fine but 253 might become an issue somewhere down the road. We could encourage short names for dependencies and drop the `-instance` suffix.

```

> As you can see, this closely mimics the `kudo install` CLI command [options](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/cmd/install.go#L56) because at the end the latter will be executed to install the operator. We omit `parameters` and `parameterFile` options at this point as they are discussed in detail [below](#dependencies-parametrization).

Author: Interesting that you and Gerred insist on this notation. My dictionary and wiki suggest the other one: https://github.com/kudobuilder/kudo/pull/1454#discussion_r412125203

1. Build a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) by transitively expanding top-level `deploy` plan using operator-tasks as vertices, and their execution order (`a` needs `b` to be installed first) as edges
2. Perform cycle detection and fail if circular dependencies found. We could additionally run this check on the client-side as part of the `kudo package verify` command to improve the UX
3. If we haven't found any cycles, start executing the top-level `deploy` plan. When encountering an operator-task, apply the corresponding `Instance` resource. Here it is the same as for any other resource that we create: we check if it is healthy and if not, end current plan execution and "wait" for it to become healthy. KUDO manager already has a [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented.

Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:

```text
AA
├── BB
│   ├── EE
│   │   ├── H
│   │   └── I
│   ├── F
│   └── GG
│       ├── J
│       └── K
├── CC
│   ├── L
│   └── M
└── D

> 

Reviewer: I think the current goal is to not allow dup dependencies - You can't use `CC` from `AA` and `BB`. You would have to define `CC1` and `CC2`, so the same operator, but different instances.

> 

Author: Exactly ^^ Life-cycle dependency is a non-goal for now.

Legend:
- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
- other tasks are marked with single letters e.g. 'D'
- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
```

In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.

> The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.

Reviewer 4: One thing that Terraform does that I really like is add a root node to that dependency graph as the "permanent" start point. This is handy so that you can have multiple "roots" (still non-cyclical!) that you can execute in parallel. Do we want to include that concept here?

Author: In the model above, the one root is the `AA` `deploy` plan. However, it becomes interesting if it has multiple _parallel_ phases, each then having an operator-task. In this case, we might actually build _multiple_ graphs and execute them in parallel. I'll keep that in the back of my head and revisit once we get to the implementation. Good point 👍

We would additionally add the higher-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of its direct children `Instance`s (e.g. `BB` and `CC`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).

The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.

> Note that in the above example if e.g. `EE` and `CC` task-operators reference the same operator package they must use distinct instance names `spec.instanceName` so that two separate `Instance`s are deployed. Otherwise, the dependency graph will have a cycle.

Author: You're right, if `EE` and `CC` were the same operator it wouldn't create a cycle. But if they both reference  **the same instance** it would be a life-cycle dependency. Which would bring us in the life-cycle dependency "hell", like what if both are configured differently, or what if `EE` is part of some other operator altogether? The solution today is to create two separate instances `EE-instance` and `CC-instance`.

> ##### Dependencies Parametrization

Reviewer 4: ```suggestion
##### Dependencies Parameterization
```

Author: Seem that both notations exists, but wiki has [Parametrization](https://en.wikipedia.org/wiki/Parametrization) as the main one.

> We want to encourage operator composition by providing a way of operator encapsulation. In other words, operator users should not be allowed to arbitrarily modify the parameters of embedded operator instances. The higher-level operator should define all parameters that its **direct** dependency  operators need. Let's demonstrate this on an example of a simple operator `AA` that has operator `BB` as a dependency. 

Reviewer: Do we plan to restrict the modification of parameters of embedded operators after the initial deployment? i.e. if `BB` was deployed as part of `AA`, can a user use `kudo update --instance BB -p PASSWORD=xyz` or will we detect that `BB` is part of `AA` and prevent that?

Reviewer 3: I don't think there is much point doing that, since the user can always `kudoctl edit instance BB` behind our back?

Reviewer: No, if that restriction is in the admission webhook they can't do that.

Reviewer 3: Hm, for the admission webhook to be able to prevent that wouldn't it have to (each time) fetch the original parent instance, and then find and re-execute the relevant `Operator` step to make sure its output matches what the client POSTed? :thinking: 
That strikes me as infeasisble.

Reviewer: Yeah, maybe. I don't think it's really a blocker. I think it would be nice to have it in the admission webhook, if it's easily implementable, but if it's too much work or too complicated we can skip it.

Author: I think this might be possible to prevent, but I'm not planning on doing that at this point. We've discussed this point with @kensipe and @porridge and there is a lot of value in allowing users to execute plans/update parameters on underlying instances in cases where some valuable features are not exposed at the top level.

```yaml
AA
└── BB
```
Operator `BB` has a required and empty parameter `PASSWORD`. To provide a way for the `AA` operator user to set the password we extend the operator-task with a new field `parameterFile`:

```yaml
tasks:
- name: deploy-bb
  kind: Operator
  spec:

>     parameterFile: bb-params.yaml # optional, defines the parameter that will be set on the bb-instance  

Reviewer 5: this seems overly complicated... it defines in the operator a static configuration by the operator dev for which the operator user would need to have deep insight into... it would be required of them to look at source to know what to do.

wouldn't a better approach be to scope parameters.... so `-p PASSWORD=foo` is what we have today...  perhaps that is just shorthand for `zookeeper:PASSWORD`.  If we need to allow for the same operator to be used for multiple tasks than perhaps we need a task id for the operator task

```yaml
tasks:
- name: deploy-bb
  id: foo
  kind: Operator
```
or we could use the name of the task (but I love the default to be the operator if we can / makes sense)

this would allow for `-p kafka:PASSWORD=foo -p zookeeper:PASSWORD=foo`

Reviewer: I like the idea of prefixes. It might even be possible to adapt that to parameter files:
`--parameter-file zookeeper:zk-params.yaml --parameter-file kafka:kafka-params.yaml` and each of the parameters in that file would get the prefix automatically.

If we have multiple operators of the same type, we might need the full graph for the prefix: `-p AA:BB:PASSWORD=foo`

Reviewer 3: > this seems overly complicated... it defines in the operator a static configuration by the operator dev for which the operator user would need to have deep insight into... it would be required of them to look at source to know what to do.

I think it's the opposite. If some app has a child RDBMS operator, it's an implementation detail of that app operator whether the RDBMS used is mysql or postgres. 
 It is up to its operator developer to:
- decide how much of the implementation to expose to end users, and in what way,
- document them, and how they affect nested operators, if any.

That way, the user of the top-level operator only has access to and needs to know about the parameters of the top-level operator. In the ideal case, 

IMHO forcing the user to know about how to configure a nested operator in detail, is creating complexity both on the user, and complicating matters for the top-level operator developer who will be unable to seamlessly switch to another operator once this detail has been leaked.

Also, I think @zen-dog already tried the prefix way, and that ended up with a [very complex implementation](https://github.com/kudobuilder/kudo/pull/1454#commits-pushed-1e200e4).

Author: As mentioned above `AA:BB:PASSWORD=foo` breaks the encapsulation and exposes the underlying implementation detail and _path to it_ to the end-user. This makes it much harder to treat e.g. underlying DB as an implementation detail. Also, the `EE` might decide to hard-code the password because it makes sense in that scenario and this detail might not be exposed at all the higher-level. 
In addition to the encapsulation argument, the proposed solution has the least amount of additional complexity in terms of implementation. I propose we start with it and reiterate if/when necessary.

```bash
$ kubectl kudo install AA -p BB_PASSWORD=secret
```  

which will create `OperatorVersion-AA` 

```yaml
# /apis/kudo.dev/v1beta1/namespaces/default/operatorversions/aa-0.1.0
spec:
  operator:
    kind: Operator
    name: dummy
  parameters:
  - name: BB_PASSWORD
    displayName: "BB password"
    description: "password for the underlying instance of BB"
    required: true
  tasks:
  - name: deploy-bb
    kind: Operator
    spec:
      parameterFile: bb-params.yaml
  templates:
    bb-params.yaml: |
      PASSWORD: {{ .Params.BB_PASSWORD }}
  plans:
    deploy:
      ...  
```

and `Instance-AA` resources

```yaml
# /apis/kudo.dev/v1beta1/namespaces/default/instances/instance-aa
spec:
  parameters:
    BB_PASSWORD: secret
```

During the execution of the `deploy-bb` task, the `bb-params.yaml` is expanded the same way we expand templates during the apply-task execution. The `deploy-bb` operator-task then creates the `Instance-BB` resource and saves the expanded parameter `PASSWORD: secret` in it.

What happens if we have a deeper nested operator-tasks tree e.g.:
```yaml
AA
└── BB
    └── CC
        └── DD  
            └── EE
```

> and it is the low-level `EE` operator that needs the password? It is like the dependency injection through constructor parameters: every higher-level operator has to encapsulate the password parameter so that `AA` has the `BB_PASSWORD`, `BB` the `CC_PASSWORD` and so on.

Reviewer: This means that (assuming that all operators require a password), that AA needs `BB_PASSWORD`, `CC_PASSWORD`, `DD_PASSWORD` and `EE_PASSWORD`, i.e. it would require a top-level-operator to repeat every parameter that is exposed by any of the dependencies. 

This seems to be not really sustainable long term - we already have operators with > 50 Parameters.

I think there is an argument to be made that not all parameters need to be exposed on the top-level, and it might be good to highlight specific parameters that *must* be set, or to prevent certain parameters of dependencies to be set, but I think it should still be possible to customise most of them.

Additionally, this would require a lot of updates on the top-level operator each time a nested operator adds new parameters.

Reviewer 3: > I think there is an argument to be made that not all parameters need to be exposed on the top-level, and it might be good to highlight specific parameters that must be set, or to prevent certain parameters of dependencies to be set, but I think it should still be possible to customise most of them.

I initially thought so as well, but from the customer feedback so far (@akirillov ) it seems like operator devs lean towards only exposing as little as necessary, rather than allowing users to tweak every little knob seven levels deep in the stack. Ensuring that every combination of 100 parameters works is impossible.

Reviewer: I would argue that this depends very much on the top-level operator. In his specific case this might be true, but for an operator that can be scaled from a single-instance to a multi-datacenter deployment a lot more knobs might need to be tunable.

And yes - ensuring that all combinations work is impossible, that is true. It still might be required to adjust one or two of those 100 parameters seven levels deep in the stack.

Reviewer 3: > I would argue that this depends very much on the top-level operator.

Right, my gut feeling is that the more complex and mature the system is, the shorter the paths will be along which a parameter is forwarded between operators unchanged.

Author: > And yes - ensuring that all combinations work is impossible, that is true. It still might be required to adjust one or two of those 100 parameters seven levels deep in the stack.

In this case we still have the underlying instances available so the user can always `kudo update --instance` them individually.

### Risks and Mitigation

The biggest risk is the increased complexity of the instance controller and the workflow engine. With the above approach, we can reuse much of the code and UX we have currently: plans and phases for flow control, local operators and custom operator repositories for easier development and deployment, and usual status reporting for debugging. The API footprint remains small as the only new API element is the [operator-task](#operator-task). Dependency graph building and traversal will require a graph library and there are a [few](https://github.com/yourbasic/graph) [out](https://godoc.org/github.com/twmb/algoimpl/go/graph) [there](https://godoc.org/gonum.org/v1/gonum/graph) so this will help mitigate some complexity.

## Alternatives

> One alternative is to use terraform and the existing [KUDO terraform provider](https://kudo.dev/blog/blog-2020-02-07-kudo-terraform-provider-1.html#current-process) to outsource the burden of dealing with the dependency graphs. On the upside, we would avoid the additional implementation complexity in KUDO _itself_ (though the complexity of the terraform provider is not going anywhere) and get [output values](https://www.terraform.io/docs/configuration/outputs.html) and [resource referencing](https://www.terraform.io/docs/configuration/resources.html#referring-to-instances) on top. On the downside, terraform is a heavy dependency which will completely replace KUDO UI. It is hard to quantify the pros and cons of both approaches, so it is left up for discussion.

Reviewer 4: I would lean _against_ using Terraform for this. Semantically, KUDO operators are their own "concept". The KUDO Terraform Provider is still very valuable even with a built-in dependency engine, as it allows KUDO to be a first class concept if a user is powering their IaC with Terraform.

However, there are others doing IaC with tools such as Pulumi, Cluster API, Crossplane, and others. Treating KUDO semantics at a separate level allows those semantics to be exposed as a single unit to any prevailing tool rather than tying a critical, core feature to TF specifics.
